File: src/blueprint/project.py
"""Stateless starter-project behavior."""

from __future__ import annotations

from dataclasses import dataclass

from .config import DEFAULT_CONFIG, ProjectConfig, define_config
from .version import __version__


@dataclass(frozen=True, slots=True)
class ProjectInfo:
    """Describe one configured starter project."""

    name: str
    version: str
    output: str


def get_project_info(config: ProjectConfig = DEFAULT_CONFIG) -> ProjectInfo:
    """Validate configuration and return immutable project information.

    Args:
        config: Complete project configuration.

    Returns:
        Immutable project information.

    Raises:
        TypeError: A field is invalid.
    """

    admitted = define_config(
        {
            "project": {"name": config.project.name},
            "cli": {"output": config.cli.output},
        }
    )
    return ProjectInfo(name=admitted.project.name, version=__version__, output=admitted.cli.output)

File: src/project.ts
import packageMetadata from "../package.json" with { type: "json" };

import { DEFAULT_CONFIG, defineConfig, type ProjectConfig } from "./config.js";

/** Blueprint's cross-branch project version. */
export const VERSION = packageMetadata.blueprintVersion;

/** Stable information about one configured starter project. */
export interface ProjectInfo {
  readonly name: string;
  readonly version: string;
  readonly output: ProjectConfig["cli"]["output"];
}

/** Validate one configuration value and return immutable project information. */
export function getProjectInfo(config: Readonly<ProjectConfig> = DEFAULT_CONFIG): Readonly<ProjectInfo> {
  const admitted = defineConfig(config);
  return Object.freeze({
    name: admitted.project.name,
    version: VERSION,
    output: admitted.cli.output,
  });
}
